File: frontend/src/stores/hosts.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'

export interface Host {
  id: number
  ip: string
  hostname: string | null
  mac: string | null
  vendor: string | null
  os: string | null
  status: string
  label: string | null
  notes: string | null
  tags: string[]
  monitoring_enabled: boolean
  first_seen: string
  last_seen: string
}

export const useHostsStore = defineStore('hosts', () => {
  const hosts = ref<Host[]>([])
  const loading = ref(false)

  async function fetchHosts() {
    loading.value = true
    const resp = await fetch('/api/v1/hosts')
    if (resp.ok) hosts.value = await resp.json()
    loading.value = false
  }

  async function updateHost(id: number, updates: Partial<Host>) {
    const resp = await fetch(`/api/v1/hosts/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updates),
    })
    if (resp.ok) {
      const updated = await resp.json()
      const idx = hosts.value.findIndex(h => h.id === id)
      if (idx !== -1) hosts.value[idx] = updated
    }
  }

  async function deleteHost(id: number) {
    await fetch(`/api/v1/hosts/${id}`, { method: 'DELETE' })
    hosts.value = hosts.value.filter(h => h.id !== id)
  }

  return { hosts, loading, fetchHosts, updateHost, deleteHost }
})

File: backend/api/hosts.py
import json
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.models import Host, Service

router = APIRouter(prefix="/hosts", tags=["hosts"])


class HostUpdate(BaseModel):
    label: str | None = None
    notes: str | None = None
    tags: list[str] | None = None
    monitoring_enabled: bool | None = None
    os: str | None = None


def host_to_dict(h: Host) -> dict:
    return {
        "id": h.id,
        "ip": h.ip,
        "hostname": h.hostname,
        "mac": h.mac,
        "vendor": h.vendor,
        "os": h.os,
        "status": h.status,
        "label": h.label,
        "notes": h.notes,
        "tags": json.loads(h.tags) if h.tags else [],
        "monitoring_enabled": h.monitoring_enabled,
        "first_seen": h.first_seen.isoformat(),
        "last_seen": h.last_seen.isoformat(),
    }


@router.get("")
def list_hosts(db: Session = Depends(get_db)) -> list[dict]:
    return [host_to_dict(h) for h in db.query(Host).order_by(Host.ip).all()]


@router.get("/{host_id}")
def get_host(host_id: int, db: Session = Depends(get_db)) -> dict:
    h = db.query(Host).filter_by(id=host_id).first()
    if not h:
        raise HTTPException(status_code=404, detail="Host not found")
    return host_to_dict(h)


@router.put("/{host_id}")
def update_host(host_id: int, body: HostUpdate, db: Session = Depends(get_db)) -> dict:
    h = db.query(Host).filter_by(id=host_id).first()
    if not h:
        raise HTTPException(status_code=404, detail="Host not found")
    if body.label is not None:
        h.label = body.label
    if body.notes is not None:
        h.notes = body.notes
    if body.tags is not None:
        h.tags = json.dumps(body.tags)
    if body.monitoring_enabled is not None:
        h.monitoring_enabled = body.monitoring_enabled
    if body.os is not None:
        h.os = body.os
    db.commit()
    db.refresh(h)
    return host_to_dict(h)


@router.delete("/{host_id}", status_code=204)
def delete_host(host_id: int, db: Session = Depends(get_db)) -> None:
    h = db.query(Host).filter_by(id=host_id).first()
    if not h:
        raise HTTPException(status_code=404, detail="Host not found")
    db.delete(h)
    db.commit()


@router.get("/{host_id}/services")
def list_services(host_id: int, db: Session = Depends(get_db)) -> list[dict]:
    h = db.query(Host).filter_by(id=host_id).first()
    if not h:
        raise HTTPException(status_code=404, detail="Host not found")
    return [
        {
            "id": s.id,
            "port": s.port,
            "protocol": s.protocol,
            "service_name": s.service_name,
            "version": s.version,
            "state": s.state,
            "last_seen": s.last_seen.isoformat() if s.last_seen else None,
        }
        for s in h.services
    ]
